Define the new Scale type in the autoscaling API group.

madhusudancs opened the pull request:
This new Scale type supports the more powerful set-based label selector
semantics. The selector, however, is stored in a serialized format, as
a string.

cc @lavalamp @bgrant0607 @piosz @davidopp @caesarxuchao @fgrzadkowski

bgrant0607 commented:
Thanks. Will take a look.

```
# k8s.io/kubernetes/contrib/mesos/pkg/controllermanager
/go/src/k8s.io/kubernetes/contrib/mesos/pkg/controllermanager/controllermanager.go:217: cannot use hpaClient.Extensions() (type "k8s.io/kubernetes/pkg/client/typed/generated/extensions/unversioned".ExtensionsInterface) as type "k8s.io/kubernetes/pkg/client/typed/generated/autoscaling/unversioned".ScalesGetter in argument to podautoscaler.NewHorizontalController:
	"k8s.io/kubernetes/pkg/client/typed/generated/extensions/unversioned".ExtensionsInterface does not implement "k8s.io/kubernetes/pkg/client/typed/generated/autoscaling/unversioned".ScalesGetter (wrong type for Scales method)
		have Scales(string) "k8s.io/kubernetes/pkg/client/typed/generated/extensions/unversioned".ScaleInterface
		want Scales(string) "k8s.io/kubernetes/pkg/client/typed/generated/autoscaling/unversioned".ScaleInterface
godep: go exit status 2
```

piosz commented:
This PR only introduces the new version of Scale in autoscaling group and doesn't touch extensions group at all. Not sure which attitude is preferred here, but I think the PR requires one of:
1. if we want to follow the same pattern that in HPA GA (see https://github.com/kubernetes/kubernetes/pull/20501#issuecomment-180059994) HPA still should use Scale in version v1beta1
2. if we want to make a breaking change, Scale should be removed from extensions group

Personally I think that 1. is preferred to satisfy what was described in https://github.com/kubernetes/kubernetes/pull/20501#issuecomment-180483743. @bgrant0607 WDYT?

lavalamp commented:
I think I agree w/ @piosz. Although we should *not* make a breaking change (unless HPA is the only user, AND we don't care about rolling apiserver back without rolling back controller-manager).

lavalamp commented:
That said, the types and implementation can come separately, I think.

bgrant0607 commented:
@piosz @lavalamp 

ReplicationController used the v1beta1 scale in 1.1, and there are non-HPA clients using it (e.g., kubectl/oc).

The fundamental problem is that the v1beta1 Scale cannot express the full set of pods that can be managed by Deployment or ReplicaSet. It's not expressive enough.

1.2 breaks compatibility for Deployment in lots of ways, so we could introduce a new, incompatible scale for it.

ReplicaSet didn't exist in 1.1, so we could also use the new scale.

How bad would it be for HPA and kubectl to use the extensions/v1beta1 scale with ReplicationController for a release or two, but to use the autoscaling/v1 scale for Deployment and ReplicaSet?

bgrant0607 commented:
I take it back: ReplicationController probably needs to use DeprecatedScale forever.

bgrant0607 commented:
Yet another possible solution would be to add both old and new selector fields to Scale, deprecate the existing field, leave the existing field nil in the case of Deployment or ReplicaSet, and remove the old field when we remove the v1beta1 HPA.

The new selector field would need to be called something different, obviously, like targetSelector or selectorQuery (since it's what would be passed via the query parameter).

madhusudancs commented:
@lavalamp @piosz In addition to what @bgrant0607 already explained:

> 1. if we want to follow the same pattern that in HPA GA (see #20501 (comment)) HPA still should use Scale in version v1beta1

We can't do that as is because of the problem that @bgrant0607 described -full set of pods cannot be described by v1beta1 Scale. We can however do that with the second solution that @bgrant0607 proposed. Having two selector types in the same API type will look a little strange, but we could do that if that's what we want. IMO, introducing autoscale/v1 Scale is cleaner.

> 2. if we want to make a breaking change, Scale should be removed from extensions group

We won't break the existing Scale outside HPA. As described here - https://github.com/kubernetes/kubernetes/issues/18528#issuecomment-176995917, extensions/v1beta1/...replicationcontrollers/.../scale will continue to serve the existing v1beta1 Scale.

HPA, Deployments and ReplicaSets will use the new autoscaling/v1 Scale. So we will need both scale types and hence extensions/v1beta1 Scale cannot be removed.

Does that make sense?

lavalamp commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -261,7 +261,7 @@ func StartControllers(s *options.CMServer, kubeClient *client.Client, kubeconfig
 				metrics.DefaultHeapsterPort,
 			)
-			podautoscaler.NewHorizontalController(hpaClient.Core(), hpaClient.Extensions(), hpaClient, metricsClient).
+			podautoscaler.NewHorizontalController(hpaClient.Core(), hpaClient.Autoscaling(), hpaClient, metricsClient).
```
This is problematic. We decided not to do this as then a new controller manager won't work with an old apiserver.

madhusudancs replied:
What's the proposed solution/alternative here?

lavalamp commented on pkg/client/typed/generated/autoscaling/unversioned/scale_expansion.go:
```diff
@@ -0,0 +1,65 @@
+)
+
+// The ScaleExpansion interface allows manually adding extra methods to the ScaleInterface.
+type ScaleExpansion interface {
```
Shouldn't this expansion go in all of the resources that have scale as a subresource, not here? AFAIK, Scale doesn't stand alone and we shouldn't be making a client for it.

caesarxuchao replied:
The clientset currently includes client for scale just to mimic what the unversioned client provides: https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/scale.go

I agree with @lavalamp that we should treat it at a subresource, like how we treat `status`. I'm just not sure if the apiserver expects the `scale` to be encoded with apiVersion=`extensions/v1beta1`, if not, this should be easy to fix.

lavalamp commented on pkg/controller/podautoscaler/horizontal_test.go:
```diff
@@ -116,17 +117,17 @@ func (tc *testCase) prepareTestClient(t *testing.T) *fake.Clientset {
 
 	fakeClient.AddReactor("get", "replicationController", func(action core.Action) (handled bool, ret runtime.Object, err error) {
-		obj := &extensions.Scale{
+		obj := &autoscaling.Scale{
```
We should not be changing the type of replicationController's scale subresource.

bgrant0607 replied:
I agree.

lavalamp commented on pkg/controller/podautoscaler/metrics/metrics_client.go:
```diff
@@ -44,11 +44,11 @@ type MetricsClient interface {
 	// (e.g. 70 means that an average pod uses 70% of the requested CPU)
 	// and the time of generation of the oldest of utilization reports for pods.
-	GetCPUUtilization(namespace string, selector map[string]string) (*int, time.Time, error)
+	GetCPUUtilization(namespace string, selector string) (*int, time.Time, error)
```
We may need to split all this code into old and new versions. :/

bgrant0607 replied:
The old selector could be converted to the string form internally, to share the implementation.

lavalamp commented:
This PR is missing the bit that actually adds the new scale subresource anywhere in apiserver? I think that has to come before the client changes.

lavalamp commented:
Or to rephrase, if you remove the client changes, then this can go in.

Then we need a PR that implements this on the server.

Then we either need a PR that lets clients use both old & new locations, OR we need to wait a release and switch the client over--that's more risky, though.

bgrant0607 commented:
I think HPA can figure out which scale resource it is by doing a GET and looking at the apiVersion. We shouldn't hardcode it by parent resource type (ReplicationController, ReplicaSet, Deployment), because that wouldn't cover Openshift usage and potential future changes.

madhusudancs committed:
Define the new Scale type in the autoscaling API group.

This new Scale type supports the more powerful set-based label selector
semantics. The selector, however, is stored in a serialized format, as
a string.

madhusudancs committed:
Generated code.

bgrant0607 commented:
LGTM

bgrant0607 commented:
Unit tests failed.

madhusudancs committed:
Teach serde test to deal with autoscaling group's duality.
